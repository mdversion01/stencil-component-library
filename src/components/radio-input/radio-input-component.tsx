import { Component, Prop, h, State, Event, EventEmitter } from '@stencil/core';

@Component({
  tag: 'radio-input-component',
  // styleUrls: ['../checkbox/checkbox.scss', './radio-input.scss', '../form-styles.scss'],
  shadow: false,
})
export class RadioComponent {
  @Prop() radio = false;
  @Prop() radioGroup = false;
  @Prop() customRadio = false;
  @Prop() customRadioGroup = false;

  @Prop() groupOptions: any = []; // JSON string or array
  @Prop() groupTitle = '';
  @Prop() groupTitleSize = '';
  @Prop() inline = false;
  @Prop() inputId = '';
  @Prop() name = '';
  @Prop() labelTxt = '';
  @Prop() required = false;
  @Prop() size = '';
  @Prop() disabled = false;
  @Prop() validation = false;
  @Prop() validationMsg = '';
  @Prop() value = '';

  @State() parsedOptions: Array<any> = [];
  @State() selectedValue = '';
  @State() singleChecked = false;

  @Event() groupChange: EventEmitter<string>;

  componentWillLoad() {
    if (Array.isArray(this.groupOptions)) {
      this.parsedOptions = this.groupOptions;
    } else if (typeof this.groupOptions === 'string') {
      try {
        this.parsedOptions = JSON.parse(this.groupOptions);
      } catch (e) {
        console.error('Invalid groupOptions JSON:', e);
        this.parsedOptions = [];
      }
    }

    const preSelected = this.parsedOptions.find(opt => opt.checked);
    if (preSelected) {
      this.selectedValue = preSelected.value;
    }
  }

  private handleGroupChange(event: Event, value: string) {
    const target = event.target as HTMLInputElement;

    if (target.checked) {
      this.selectedValue = value;

      const updated = this.parsedOptions.map(opt => ({
        ...opt,
        checked: opt.value === value,
      }));
      this.parsedOptions = updated;

      this.groupChange.emit(this.selectedValue);
    }
  }

  private handleSingleChange(event: Event) {
    const target = event.target as HTMLInputElement;
    this.singleChecked = target.checked;
  }

  render() {
    // GROUP RENDER
    if (this.radioGroup || this.customRadioGroup) {
      const wrapperClass = this.customRadioGroup ? 'custom-control custom-radio' : 'form-check';
      const inputClass = this.customRadioGroup ? 'custom-control-input' : 'form-check-input';
      const labelClass = this.customRadioGroup ? 'custom-control-label' : 'form-check-label';

      const showValidation = this.validation && this.required && !this.selectedValue;

      return (
        <div class={`radio-group ${showValidation ? 'was-validated' : ''}`}>
          <div class={`group-title ${this.groupTitleSize}`}>
            {this.groupTitle}
            {this.required ? <span class="required">*</span> : ''}
          </div>
          <div class={`form-group ${this.inline ? 'form-inline' : ''}`}>
            {this.parsedOptions.map(option => (
              <div class={wrapperClass}>
                <input
                  class={inputClass}
                  type="radio"
                  name={this.name}
                  id={option.inputId}
                  value={option.value}
                  checked={option.checked}
                  disabled={option.disabled}
                  required={this.required}
                  aria-checked={option.checked ? 'true' : 'false'}
                  aria-disabled={option.disabled ? 'true' : 'false'}
                  onChange={e => this.handleGroupChange(e, option.value)}
                />
                <label class={`${labelClass} ${this.size}`} htmlFor={option.inputId}>
                  {option.labelTxt}
                </label>
              </div>
            ))}
          </div>
          {showValidation && this.validationMsg ? (
            <div class="invalid-feedback form-text" style={{ display: showValidation ? 'block' : '' }}>{this.validationMsg}</div>
          ) : null}
        </div>
      );
    }

    // SINGLE RENDER
    const singleWrapperClass = this.customRadio ? 'custom-control custom-radio' : 'form-check';
    const singleInputClass = this.customRadio ? 'custom-control-input' : 'form-check-input';
    const singleLabelClass = this.customRadio ? 'custom-control-label' : 'form-check-label';

    const showSingleValidation = this.validation && this.required && !this.singleChecked;

    return (
      <div class={`form-group ${showSingleValidation ? 'was-validated' : ''}`}>
        <div class={singleWrapperClass}>
          <input
            class={singleInputClass}
            id={this.inputId}
            type="radio"
            name={this.name}
            value={this.value}
            disabled={this.disabled}
            required={this.required}
            aria-checked={this.singleChecked ? 'true' : 'false'}
            aria-disabled={this.disabled ? 'true' : 'false'}
            onChange={e => this.handleSingleChange(e)}
          />
          <label class={`${singleLabelClass} ${this.size}`} htmlFor={this.inputId}>
            {this.labelTxt}
            {this.required ? <span class="required">*</span> : ''}
          </label>
          {showSingleValidation && this.validationMsg ? (
            <div class="invalid-feedback form-text">{this.validationMsg}</div>
          ) : null}
        </div>
      </div>
    );
  }
}
