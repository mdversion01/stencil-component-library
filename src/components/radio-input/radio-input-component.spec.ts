import { newSpecPage } from '@stencil/core/testing';
import { RadioComponent } from './radio-input-component';

describe('radio-input-component', () => {
  it('renders single radio input', async () => {
    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component radio input-id="single-radio" name="test" label-txt="Single Radio"></radio-input-component>`,
    });
    const input = page.root.querySelector('input');
    expect(input.getAttribute('type')).toBe('radio');
    expect(input.getAttribute('name')).toBe('test');
    expect(page.root).toMatchSnapshot();
  });

  it('renders single custom radio input', async () => {
    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component custom-radio input-id="custom-radio" name="custom" label-txt="Custom Radio"></radio-input-component>`,
    });
    expect(page.root.querySelector('.custom-control')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('renders radio group with options', async () => {
    const options = JSON.stringify([
      { value: 'one', labelTxt: 'Option One', inputId: 'radio-one', checked: false, disabled: false },
      { value: 'two', labelTxt: 'Option Two', inputId: 'radio-two', checked: true, disabled: false },
      { value: 'three', labelTxt: 'Option Three', inputId: 'radio-three', checked: false, disabled: true },
    ]);

    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component radio-group name="group" group-title="Select Option" group-options='${options}'></radio-input-component>`,
    });

    const inputs = page.root.querySelectorAll('input[type="radio"]');
    expect(inputs.length).toBe(3);
    expect(inputs[1].checked).toBe(true);
    expect(inputs[2].disabled).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('renders custom radio group with options', async () => {
    const options = JSON.stringify([
      { value: 'alpha', labelTxt: 'Alpha', inputId: 'radio-alpha', checked: false, disabled: false },
      { value: 'beta', labelTxt: 'Beta', inputId: 'radio-beta', checked: false, disabled: false },
    ]);

    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component custom-radio-group name="customGroup" group-title="Custom Group" group-options='${options}'></radio-input-component>`,
    });

    expect(page.root.querySelectorAll('input[type="radio"]').length).toBe(2);
    expect(page.root.querySelector('.custom-control')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('shows validation message when no selection is made and required', async () => {
    const options = JSON.stringify([
      { value: 'x', labelTxt: 'X', inputId: 'radio-x', checked: false, disabled: false },
      { value: 'y', labelTxt: 'Y', inputId: 'radio-y', checked: false, disabled: false },
    ]);

    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component radio-group name="group1" group-title="Validation Test" required validation validation-msg="Please select one" group-options='${options}'></radio-input-component>`,
    });

    expect(page.root.querySelector('.invalid-feedback')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('removes validation message when an option is selected', async () => {
    const options = JSON.stringify([
      { value: 'x', labelTxt: 'X', inputId: 'radio-x', checked: false, disabled: false },
      { value: 'y', labelTxt: 'Y', inputId: 'radio-y', checked: false, disabled: false },
    ]);

    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component radio-group name="group2" group-title="Validation Clear" required validation validation-msg="Please select one" group-options='${options}'></radio-input-component>`,
    });

    const inputs = page.root.querySelectorAll('input[type="radio"]');
    inputs[0].checked = true;
    inputs[0].dispatchEvent(new Event('change', { bubbles: true }));
    await page.waitForChanges();

    expect(page.root.querySelector('.invalid-feedback')).toBeFalsy();
    expect(page.root).toMatchSnapshot();
  });

  it('emits groupChange event when selection changes', async () => {
    const options = JSON.stringify([
      { value: 'apple', labelTxt: 'Apple', inputId: 'radio-apple', checked: false, disabled: false },
      { value: 'banana', labelTxt: 'Banana', inputId: 'radio-banana', checked: false, disabled: false },
    ]);

    const page = await newSpecPage({
      components: [RadioComponent],
      html: `<radio-input-component radio-group name="fruits" group-title="Choose Fruit" group-options='${options}'></radio-input-component>`,
    });

    const spy = jest.fn();
    page.root.addEventListener('groupChange', spy);

    const inputs = page.root.querySelectorAll('input[type="radio"]');
    inputs[0].checked = true;
    inputs[0].dispatchEvent(new Event('change', { bubbles: true }));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalled();
  });
});
