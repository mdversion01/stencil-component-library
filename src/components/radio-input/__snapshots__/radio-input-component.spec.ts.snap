// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio-input-component removes validation message when an option is selected 1`] = `
<radio-input-component group-options="[{&quot;value&quot;:&quot;x&quot;,&quot;labelTxt&quot;:&quot;X&quot;,&quot;inputId&quot;:&quot;radio-x&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false},{&quot;value&quot;:&quot;y&quot;,&quot;labelTxt&quot;:&quot;Y&quot;,&quot;inputId&quot;:&quot;radio-y&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false}]" group-title="Validation Clear" name="group2" radio-group="" required="" validation="" validation-msg="Please select one">
  <div class="radio-group">
    <div class="group-title">
      Validation Clear
      <span class="required">
        *
      </span>
    </div>
    <div class="form-group">
      <div class="form-check">
        <input aria-checked="true" aria-disabled="false" checked="" class="form-check-input" id="radio-x" name="group2" required="" type="radio" value="x">
        <label class="form-check-label" htmlfor="radio-x">
          X
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="radio-y" name="group2" required="" type="radio" value="y">
        <label class="form-check-label" htmlfor="radio-y">
          Y
        </label>
      </div>
    </div>
  </div>
</radio-input-component>
`;

exports[`radio-input-component renders custom radio group with options 1`] = `
<radio-input-component custom-radio-group="" group-options="[{&quot;value&quot;:&quot;alpha&quot;,&quot;labelTxt&quot;:&quot;Alpha&quot;,&quot;inputId&quot;:&quot;radio-alpha&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false},{&quot;value&quot;:&quot;beta&quot;,&quot;labelTxt&quot;:&quot;Beta&quot;,&quot;inputId&quot;:&quot;radio-beta&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false}]" group-title="Custom Group" name="customGroup">
  <div class="radio-group">
    <div class="group-title">
      Custom Group
    </div>
    <div class="form-group">
      <div class="custom-control custom-radio">
        <input aria-checked="false" aria-disabled="false" class="custom-control-input" id="radio-alpha" name="customGroup" type="radio" value="alpha">
        <label class="custom-control-label" htmlfor="radio-alpha">
          Alpha
        </label>
      </div>
      <div class="custom-control custom-radio">
        <input aria-checked="false" aria-disabled="false" class="custom-control-input" id="radio-beta" name="customGroup" type="radio" value="beta">
        <label class="custom-control-label" htmlfor="radio-beta">
          Beta
        </label>
      </div>
    </div>
  </div>
</radio-input-component>
`;

exports[`radio-input-component renders radio group with options 1`] = `
<radio-input-component group-options="[{&quot;value&quot;:&quot;one&quot;,&quot;labelTxt&quot;:&quot;Option One&quot;,&quot;inputId&quot;:&quot;radio-one&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false},{&quot;value&quot;:&quot;two&quot;,&quot;labelTxt&quot;:&quot;Option Two&quot;,&quot;inputId&quot;:&quot;radio-two&quot;,&quot;checked&quot;:true,&quot;disabled&quot;:false},{&quot;value&quot;:&quot;three&quot;,&quot;labelTxt&quot;:&quot;Option Three&quot;,&quot;inputId&quot;:&quot;radio-three&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:true}]" group-title="Select Option" name="group" radio-group="">
  <div class="radio-group">
    <div class="group-title">
      Select Option
    </div>
    <div class="form-group">
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="radio-one" name="group" type="radio" value="one">
        <label class="form-check-label" htmlfor="radio-one">
          Option One
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="true" aria-disabled="false" checked="" class="form-check-input" id="radio-two" name="group" type="radio" value="two">
        <label class="form-check-label" htmlfor="radio-two">
          Option Two
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="false" aria-disabled="true" class="form-check-input" disabled="" id="radio-three" name="group" type="radio" value="three">
        <label class="form-check-label" htmlfor="radio-three">
          Option Three
        </label>
      </div>
    </div>
  </div>
</radio-input-component>
`;

exports[`radio-input-component renders single custom radio input 1`] = `
<radio-input-component custom-radio="" input-id="custom-radio" label-txt="Custom Radio" name="custom">
  <div class="form-group">
    <div class="custom-control custom-radio">
      <input aria-checked="false" aria-disabled="false" class="custom-control-input" id="custom-radio" name="custom" type="radio" value="">
      <label class="custom-control-label" htmlfor="custom-radio">
        Custom Radio
      </label>
    </div>
  </div>
</radio-input-component>
`;

exports[`radio-input-component renders single radio input 1`] = `
<radio-input-component input-id="single-radio" label-txt="Single Radio" name="test" radio="">
  <div class="form-group">
    <div class="form-check">
      <input aria-checked="false" aria-disabled="false" class="form-check-input" id="single-radio" name="test" type="radio" value="">
      <label class="form-check-label" htmlfor="single-radio">
        Single Radio
      </label>
    </div>
  </div>
</radio-input-component>
`;

exports[`radio-input-component shows validation message when no selection is made and required 1`] = `
<radio-input-component group-options="[{&quot;value&quot;:&quot;x&quot;,&quot;labelTxt&quot;:&quot;X&quot;,&quot;inputId&quot;:&quot;radio-x&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false},{&quot;value&quot;:&quot;y&quot;,&quot;labelTxt&quot;:&quot;Y&quot;,&quot;inputId&quot;:&quot;radio-y&quot;,&quot;checked&quot;:false,&quot;disabled&quot;:false}]" group-title="Validation Test" name="group1" radio-group="" required="" validation="" validation-msg="Please select one">
  <div class="radio-group was-validated">
    <div class="group-title">
      Validation Test
      <span class="required">
        *
      </span>
    </div>
    <div class="form-group">
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="radio-x" name="group1" required="" type="radio" value="x">
        <label class="form-check-label" htmlfor="radio-x">
          X
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="radio-y" name="group1" required="" type="radio" value="y">
        <label class="form-check-label" htmlfor="radio-y">
          Y
        </label>
      </div>
    </div>
    <div class="form-text invalid-feedback" style="display: block;">
      Please select one
    </div>
  </div>
</radio-input-component>
`;
